### ISSUE TITLE
Support for multiple data types

### ISSUE THREAD

**Issue by eaplatanios** [deeplearning4j/nd4j#1690]
Hi,

I want to use ND4j for a project of mine but my understanding right now is that you cannot have an Integer NDArray and a Double NDArray at the same time. Is it true that all tensors have to be of the same data type? If yes, what prevents you from supporting multiple data types and casting when needed based on the Ops used? And, do you have any plans to add support for multiple data types in the future?

Thanks,
Anthony

**Comment by agibsonccc** [comment]
You can sorta do it but why do you want multiple data types? We handle casting everything to the proper data type for you anyways.

**Comment by eaplatanios** [comment]
How can you do it though? An example setting where it could be useful is when you a set of instances, each having a set of features (Double precision floats) and a set of labels (could be Boolean, for binary, or 8-bit unsigned integers, for example), in the context of machine learning. Then it would be wasteful to represent those labels as a tensor of doubles, which is the precision of the features.

**Comment by eaplatanios** [comment]
Numpy is an example of a library that can handle such a case and which also uses BLAS.

**Comment by raver119** [comment]
Don't get me wrong please, but type cast for array literally means this sequence:

- create new array to match new data type length
- loop through original array and convert each element to new data type
- destroy original array eventually

Are you 100% sure that your savings on 32 labels (average batch size), will be worth this hassle?

We do support data type conversion currently. But i just don't see **real** benefits here.

**Comment by eaplatanios** [comment]
I see your point but please consider the following use case:

We train a neural net using the cross entropy loss. We can implement this loss function efficiently by using int-valued labels to index on the double-valued predictions tensor (containing the probabilities for each label). I think this would be much less efficient had we been using double-valued labels, but please correct me if I'm wrong.

Note also that similar examples could be constructed for Natural Language Processing (NLP) applications where we often deal with discrete data structures.

**Comment by raver119** [comment]
No, it doesn't work that way. All BLAS functions use floating point data types, so even if you're using INT somewhere in python, it's only at your level. When things go down to BLAS - your INT value becomes discrete FLOAT value. Or DOUBLE. Or HALF. Or whatever else. 

int 1 == float 1.0 == double 1.0 == float16 1.0 
int 3 == float 3.0 == double 3.0 == float16 3.0 

and so on. That's due to simple fact: There's just no integer data types in BLAS specs. Float. Double. Complex. Now new data type gets popularity: HALF precision, aka float16, aka fp16, aka half1

**Comment by eaplatanios** [comment]
I see. So, what happens when you index a tensor with an int (i.e., to obtain the ith value of a vector)?

**Comment by agibsonccc** [comment]
You get back a view with the underlying data type being float or double.

**Comment by eaplatanios** [comment]
That's true but i (i.e., the index to the vector) was provided and used as an integer. It was never converted to a double value. So, back to my example, what if you use values of one int-valued tensor (i.e., the labels tensor) as an index to another tensor that's double-valued (i.e., the predictions tensor). It would be too inefficient in this case to also store the labels as a double-valued tensor as they're only used as indices into another tensor.

**Comment by raver119** [comment]
No. It would be inefficient to handle dtype conversion all the way through the pipeline. And converting ints to doubles/floats once - is quite ok. Lesser evil, you know :)

**Comment by eaplatanios** [comment]
I agree but my point in this case is that if both data types are supported at the same time, then no conversion would be necessary at any point. You would directly use the integer labels as indices in the predictions tensor that contains double values and you would get back another double-valued tensor.

The sparse cross entropy function in TensorFlow is implemented this way and it is much more efficient than the cross-entropy function using one-hot labels (which is what I would need to do given the current data type constraint of ND4j).

**Comment by raver119** [comment]
And what stops you from doing the same with double data type? How big values we're speaking about? Are they big enough to loose integral part precision?

**Comment by eaplatanios** [comment]
How would you index an array with a double value? You would need to cast it to an integer at some point to obtain a memory address eventually, right? So not supporting an int tensor in this case has two disadvantages: (i) more memory consumption, (ii) eventual cast to an integer.

**Comment by raver119** [comment]
1) Nobody cares about those 32 labels in double dtype, seriously. And in case of float, memory footprint will  be equal to integer datatype, 4 bytes per element.

2) I've explained that above: it's more efficient to cast once (as index), then doing type checks & occasional type casting through the whole math pipeline.

That's why everyone out there eventually have everything in the same data type with floating point values.

**Comment by eaplatanios** [comment]
It's not true that everyone out there uses floats and that there is no performance impact. The TensorFlow sparse cross-entropy example I gave above is a good example of where data types matter and you gain a lot of efficiency by using them appropriately. And the same example applies to Theano, Torch, MXNet, and most major deep learning frameworks out there.

**Comment by raver119** [comment]
1) Where you've got idea of "no performance impact"? I didn't say that. 

2) Point me to the code of op that has chance of going into BLAS pipeline with mixed data type in any framework.
